### PR TITLE
Configurable compression of content-types.

### PR DESCRIPTION
**Pull request by jacobstr**
Along for the ride are functional options to stop the combinatorial growth of option-specific factories.

### CODE REVIEW

**Comment by jacobstr** [comment]
See also: #40 and #43

File: gzip.go [hunk 1]
```diff
@@ -258,6 +274,45 @@ func NewGzipLevelAndMinSize(level, minSize int) (func(http.Handler) http.Handler
...
+	}
+
+	if c.minSize < 0 {
+		return fmt.Errorf("minimum size must be more than zero")
+	}
+
+	return nil
+}
+
+type GzipOpt func(c *GzipConfig)
```
Comment (jprobinson) [inline gzip.go]: To avoid some stutter, could we rename this type to `Option`?

File: gzip.go [hunk 2]
```diff
@@ -258,6 +274,45 @@ func NewGzipLevelAndMinSize(level, minSize int) (func(http.Handler) http.Handler
 	}, nil
 }
 
+// Used for functional configuration.
+type GzipConfig struct {
```
Comment (jprobinson) [inline gzip.go]: Does this need to be public? Also, to reduce stutter, can we just call this `config`?

File: gzip.go [hunk 3]
```diff
@@ -258,6 +274,45 @@ func NewGzipLevelAndMinSize(level, minSize int) (func(http.Handler) http.Handler
...
+	minSize      int
+	level        int
+	contentTypes []string
+}
+
+func (c *GzipConfig) Validate() error {
```
Comment (jprobinson) [inline gzip.go]: Does this need to be public?

File: gzip.go [hunk 4]
```diff
@@ -80,6 +80,8 @@ type GzipResponseWriter struct {
 
 	minSize int    // Specifed the minimum response size to gzip. If the response length is bigger than this value, it is compressed.
 	buf     []byte // Holds the first part of the write before reaching the minSize or the end of the write.
+
+	contentTypes []string // Only compress if the response is one of these content-types. All are accepted it empty.
```
Comment (darrenmcc) [inline gzip.go]: it --> if

File: gzip.go [hunk 5]
```diff
@@ -273,6 +328,22 @@ func acceptsGzip(r *http.Request) bool {
...
 
+// returns true if we've been configured to compress the specific content type.
+func handleContentType(contentTypes []string, w http.ResponseWriter) bool {
+	// If contentTypes is empty we handle all content types.
+	if len(contentTypes) == 0 {
+		return true
+	}
+
+	for _, c := range contentTypes {
+		if strings.ToLower(c) == strings.ToLower(w.Header().Get(contentType)) {
```
Comment (darrenmcc) [inline gzip.go]: perhaps we should lowercase the contentTypes once when the handler is initialized

**Comment by jacobstr** [comment]
Cheers for the quick review. Stylistic nits were quite sensible. Option type was also made private.

**Comment by jacobstr** [comment]
There's a slight caveat on what to do if the Content-Type is empty and cannot be determined. Right now you _could_ make it work by passing `[]string{""}`.

[]string{} - Match any content type (backwards compatible, default)
[]string{"","application/json"} - Match unknown content types and application/json.

That's kind of incidental to how the comparsion works given the value coming out of `Header().Get("Content-Type")`.

**Comment by jprobinson** [comment]
🐈 thanks!!

**Pull request merged**